src/sxg_sig.c

```diff
@@ -137,79 +183,168 @@ bool sxg_sig_generate_sig(const char* fallback_url, const sxg_buffer_t* header,
 
   // 9.  The 8-byte big-endian encoding of the length in bytes of
   // "responseHeaders", followed by the bytes of "responseHeaders".
-  success = success && sxg_write_int(header->size, 8, &message) &&
-            sxg_write_buffer(header, &message);
+  success = success && sxg_write_int(header_size, 8, &message) &&
+            sxg_write_bytes(header, header_size, &message);
 
   // Generate sigature of the message.
-  sxg_buffer_release(&sig->sig);
-  success = success && sxg_evp_sign(private_key, &message, &sig->sig);
+  sig->sig = OPENSSL_realloc(
+      sig->sig, sxg_evp_sign_size(private_key, message.data, message.size));
+
+  success = success && sig->sig != NULL;
+
+  sig->sig_size =
+      sxg_evp_sign(private_key, message.data, message.size, sig->sig);
+
+  success = success && sig->sig_size > 0;
   sxg_buffer_release(&message);
 
   return success;
 }
 
-static bool sxg_write_structured_header_binary(const sxg_buffer_t* binary,
-                                               sxg_buffer_t* target) {
-  return sxg_write_byte('*', target) && sxg_base64encode(binary, target) &&
-         sxg_write_byte('*', target);
+static size_t sxg_write_structured_header_binary_size(size_t size) {
+  return 1 + sxg_base64encode_size(size) + 1;
+}
+
+static size_t sxg_write_structured_header_binary(const uint8_t* binary,
+                                                 size_t length,
+                                                 uint8_t* target) {
+  size_t wrote = 0;
+  memcpy(&target[wrote++], "*", 1);
+  if (!sxg_base64encode(binary, length, &target[wrote])) {
+    return false;
+  }
+  wrote += sxg_base64encode_size(length);
+  memcpy(&target[wrote++], "*", 1);
+  return wrote;
+}
+
+static size_t sxg_write_structured_header_string_size(size_t size) {
+  return 1 + size + 1;
+}
+
+static size_t sxg_write_structured_header_string(const char* string,
+                                                 size_t length,
+                                                 uint8_t* target) {
+  size_t wrote = 0;
+  memcpy(&target[wrote++], "\"", 1);
```

Comment:
`target[wrote++]` = '\"' is more clear.  Repeated a few times.

Reply:
Done.